src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,835 @@
+        {
+            if(d != 1)
+                return false;
+            new_dims.push_back({{dimension::sub{1}}});
+        }
+    }
+    assert(rdims.size() == new_dims.size());
+    if(rdims.size() != new_dims.size())
+        return false;
+    dimensions = new_dims;
+    return true;
+}
+bool shape_transform_descriptor::apply_transpose(const std::vector<std::int64_t>& permutation)
+{
+    if(permutation.size() != dimensions.size())
+        return false;
+    dimensions = reorder_dims(dimensions, permutation);
+    return true;
+}
+
+bool shape_transform_descriptor::apply_broadcast(const std::vector<std::size_t>& out_lens,
+                                                 optional<std::size_t> axis)
+{
+    auto offset = axis.value_or(out_lens.size() - dimensions.size());
+
+    std::vector<dimension> new_dims;
+    std::transform(out_lens.begin(),
+                   out_lens.begin() + offset,
+                   std::back_inserter(new_dims),
+                   [&](auto len) -> dimension {
+                       return {{dimension::sub{len, {}}}};
+                   });
+    std::transform(dimensions.begin(),
+                   dimensions.end(),
+                   out_lens.begin() + offset,
+                   std::back_inserter(new_dims),
+                   [&](const dimension& dim, auto len) -> dimension {
+                       if(len == dim.len())
+                           return dim;
+                       if(dim.len() != 1)
+                           MIGRAPHX_THROW("Wrong out_lens for broadcast");
+                       return {{dimension::sub{len, {}}}};
+                   });
+    std::transform(out_lens.begin() + offset + dimensions.size(),
+                   out_lens.end(),
+                   std::back_inserter(new_dims),
+                   [&](auto len) -> dimension {
+                       return {{dimension::sub{len, {}}}};
+                   });
+    assert(out_lens.size() == new_dims.size());
+    dimensions = new_dims;
+    return true;
+}
+
+// Remove subdimensions of 1
+void remove_1_sub_dims(std::vector<dimension::sub>& subdimensions)
+{
+    subdimensions.erase(std::remove_if(subdimensions.begin(),
+                                       subdimensions.end(),
+                                       [&](const dimension::sub& d) { return d.len == 1; }),
+                        subdimensions.end());
+}
+
+void dimension::simplify()
+{
+    if(subdimensions.size() < 2)
+        return;
+    remove_1_sub_dims(subdimensions);
+    // Flatten adjacent dimensions
+    adjacent_for_each(subdimensions.begin(), subdimensions.end(), [&](sub& d1, sub& d2) {
+        if(d1.axis.size() < 2)
+            return;
+        if(d2.axis.size() < 2)
+            return;
+        if(not std::equal(d1.axis.begin(), d1.axis.end() - 1, d2.axis.begin(), d2.axis.end() - 1))
+            return;
+        auto a1 = d1.axis.back();
+        auto a2 = d2.axis.back();
+        if(a2 <= a1)
+            return;
```

Comment:
This strictly needs to be `a2 < a1` if `a2 == a1` then both subdimensions have exactly the same values for all  axes which seems unlikely/invalid and  i think it should throw.

Reply:
I'll add an assert for it.